src/autoPagination.ts

```diff
@@ -92,39 +78,97 @@ export function makeAutoPaginationMethods(
       );
     }
 
-    if (i < pageResult.data.length) {
-      const idx = reverseIteration ? pageResult.data.length - 1 - i : i;
+    const reverseIteration = isReverseIteration(this.requestArgs);
+    if (this.i < pageResult.data.length) {
+      const idx = reverseIteration
+        ? pageResult.data.length - 1 - this.i
+        : this.i;
       const value = pageResult.data[idx];
-      i += 1;
+      this.i += 1;
 
       return {value, done: false};
     } else if (pageResult.has_more) {
       // Reset counter, request next page, and recurse.
-      i = 0;
-      pagePromise = getNextPagePromise(pageResult);
-      return pagePromise.then(iterate);
+      this.i = 0;
+      this.pagePromise = this.getNextPage(pageResult);
+      return this.pagePromise.then((pageResult) => this.iterate(pageResult));
     }
     return {value: undefined, done: true};
   }
 
-  function asyncIteratorNext(): Promise<any> {
-    return memoizedPromise(promiseCache, (resolve, reject) => {
-      return pagePromise
-        .then(iterate)
+  getNextPage(_pageResult: PageResult<T>): Promise<PageResult<T>> {
+    throw new Error('Unimplemented');
+  }
+
+  next(): Promise<IterationResult<T>> {
+    return memoizedPromise(this.promiseCache, (resolve, reject) => {
+      return this.pagePromise
+        .then((pageResult) => this.iterate(pageResult))
```

Comment:
Same question about async/await here.

Reply:
I don't think so? It feels wrong to make the `(resolve, reject) => ...` callback async, because the type of that function (the callback accepted by memoizedPromise) is
```
(resolve: (value: T) => void, reject: (reason?: any) => void) => void
```
and I think you should only make functions async if their return type is a promise.

Higher-level, I suspect there might be a cleaner way to do all of whatever this memoizedPromise stuff is doing, but detangling that seems pretty involved and I didn't want that to be in scope for this PR.